src/lib/storage/disk_size.rb

```diff
@@ -0,0 +1,260 @@
+
+require "yast"
+require "pp"
+
+# This file can be invoked separately for minimal testing.
+
+module Yast
+  module Storage
+    #
+    # Class to handle disk sizes in the MB/GB/TB range with readable output.
+    #
+    class DiskSize
+      include Comparable
+
+      UNITS = ["kiB", "MiB", "GiB", "TiB", "PiB", "EiB", "ZiB", "YiB"]
+      UNLIMITED = "unlimited"
+
+      attr_accessor :size_k
+
+      def initialize(size_k = 0)
+        @size_k = size_k.round
+      end
+
+      #
+      # Factory methods
+      #
+      class << self
+        def kiB(size)
+          DiskSize.new(size)
+        end
+
+        def MiB(size)
+          DiskSize.new(size * 1024)
+        end
+
+        def GiB(size)
+          DiskSize.new(size * (1024**2))
+        end
+
+        def TiB(size)
+          DiskSize.new(size * (1024**3))
+        end
+
+        def PiB(size)
+          DiskSize.new(size * (1024**4))
+        end
+
+        def EiB(size)
+          DiskSize.new(size * (1024**5))
+        end
+
+        def ZiB(size)
+          DiskSize.new(size * (1024**6))
+        end
+
+        def YiB(size)
+          DiskSize.new(size * (1024**7))
+        end
+
+        def unlimited
+          DiskSize.new(-1)
+        end
+
+        def zero
+          DiskSize.new(0)
+        end
+
+        # Create a DiskSize from a parsed string.
+        # Valid formats:
+        #   42 GiB
+        #   42.00  GiB
+        #   42          (=> 42 kiB)
+        #   unlimited   (=> -1 == unlimited)
+        #
+        # Invalid:
+        #   42 GB    (supporting binary units only)
+        #
+        def parse(str)
+          str.strip!
+          return DiskSize.unlimited if str == UNLIMITED
+          size_str, unit = str.split(/\s+/)
+          raise ArgumentError, "Bad number: #{size_str}" if size_str !~ /^\d+\.?\d*$/
+          size = size_str.to_f
+          return DiskSize.new(size) if unit.nil?
+          DiskSize.new(size * unit_multiplier(unit))
+        end
+
+        alias_method :from_s, :parse
+        alias_method :from_human_readable, :parse
+
+        # Return the unit exponent for any of the known binary units ("kiB",
+        # "MiB", ...). The base of this exponent is 1024. The base unit is kiB.
+        #
+        def unit_exponent(unit)
+          index = UNITS.index(unit)
+          raise ArgumentError, "expected one of #{UNITS}" if index.nil?
+          index
+        end
+
+        # Return the unit multiplier for any of the known binary units ("kiB",
+        # "MiB", ...). The base unit is kiB.
+        #
+        def unit_multiplier(unit)
+          1024 ** unit_exponent(unit)
+        end
+      end
+
+      #
+      # Operators
+      #
+
+      def +(other)
+        if other.is_a?(Numeric)
+          DiskSize.new(@size_k + other)
+        elsif other.respond_to?(:size_k)
+          DiskSize.new(@size_k + other.size_k)
+        else
+          raise TypeError, "Numeric value or DiskSize expected"
```

Comment:
It often helps, e.g. when you see NilClass you know that something try to add nil